frcxx_nft/src/state.rs

```diff
@@ -0,0 +1,207 @@
+    pub owner_data: Cid,
+    /// The next available token id for minting
+    pub next_token: TokenID,
+    /// The number of minted tokens less the number of burned tokens
+    pub total_supply: u64,
+}
+
+// TODO: benchmark and tune these values
+const AMT_BIT_WIDTH: u32 = 5;
+const HAMT_BIT_WIDTH: u32 = 3;
+
+type Result<T> = std::result::Result<T, StateError>;
+
+impl NFTState {
+    /// Create a new NFT state-tree, without committing it (the root Cid) to a blockstore
+    pub fn new<BS: Blockstore>(store: &BS) -> Result<Self> {
+        // Blockstore is still needed to create valid Cids for the Hamts
+        let empty_token_array =
+            Amt::<TokenData, &BS>::new_with_bit_width(store, AMT_BIT_WIDTH).flush()?;
+        // Blockstore is still needed to create valid Cids for the Hamts
+        let empty_owner_map =
+            Hamt::<&BS, OwnerData, ActorID>::new_with_bit_width(store, HAMT_BIT_WIDTH).flush()?;
+
+        Ok(Self {
+            token_data: empty_token_array,
+            owner_data: empty_owner_map,
+            next_token: 0,
+            total_supply: 0,
+        })
+    }
+
+    pub fn load<BS: Blockstore>(store: &BS, root: &Cid) -> Result<Self> {
+        match store.get_cbor::<Self>(root) {
+            Ok(Some(state)) => Ok(state),
+            _ => panic!(""),
+        }
+    }
+
+    pub fn save<BS: Blockstore>(&self, store: &BS) -> Result<Cid> {
+        let serialized = match fvm_ipld_encoding::to_vec(self) {
+            Ok(s) => s,
+            Err(err) => return Err(StateError::Other(err.to_string())),
+        };
+        let block = Block { codec: DAG_CBOR, data: serialized };
+        let cid = match store.put(Code::Blake2b256, &block) {
+            Ok(cid) => cid,
+            Err(err) => return Err(StateError::Other(err.to_string())),
+        };
+        Ok(cid)
+    }
+
+    fn get_token_data_amt<'bs, BS: Blockstore>(
+        &self,
+        store: &'bs BS,
+    ) -> Result<Amt<TokenData, &'bs BS>> {
+        let res = Amt::load(&self.token_data, store)?;
+        Ok(res)
+    }
+
+    fn get_owner_data_hamt<'bs, BS: Blockstore>(
+        &self,
+        store: &'bs BS,
+    ) -> Result<Hamt<&'bs BS, OwnerData, ActorID>> {
+        let res = Hamt::load_with_bit_width(&self.owner_data, store, HAMT_BIT_WIDTH)?;
+        Ok(res)
+    }
+
+    /// Mint a new token to the specified address
+    pub fn mint_token<BS: Blockstore>(
+        &mut self,
+        bs: &BS,
+        owner: ActorID,
+        metadata_uri: String,
+    ) -> Result<TokenID> {
+        // update token data array
+        let mut token_array = self.get_token_data_amt(bs)?;
+        let token_id = self.next_token;
+        token_array.set(token_id, TokenData { owner, approved: vec![], metadata_uri })?;
+
+        // update owner data map
+        let mut owner_map = self.get_owner_data_hamt(bs)?;
+        let new_owner_data = match owner_map.delete(&owner) {
```

Comment:
I think `delete` will do some recomputation of intermediate nodes, which is a waste when we're going to put a value back at the same path.